metagpt/actions/write_code.py

```diff
@@ -127,20 +155,71 @@ async def run(self, *args, **kwargs) -> CodingContext:
         return coding_context
 
     @staticmethod
-    async def get_codes(task_doc, exclude, project_repo) -> str:
+    async def get_codes(
+        task_doc: Document, exclude: str, project_repo: ProjectRepo, mode: Literal["normal", "incremental"] = "normal"
+    ) -> str:
+        """
+        Get code snippets based on different modes.
+
+        Attributes:
+            task_doc (Document): Document object of the task file.
+            exclude (str): Specifies the filename to be excluded from the code snippets.
+            project_repo (ProjectRepo): ProjectRepo object of the project.
+            mode (str): Specifies the mode, either "normal" or "incremental" (default is "normal").
+
+        Returns:
+            str: Code snippets.
+
+        Description:
+        If mode is set to "normal", it returns code snippets for the regular coding phase,
+        i.e., all the code generated before writing the current file.
+
+        If mode is set to "incremental", it returns code snippets for generating the code plan and change,
+        building upon the existing code in the "normal" mode and adding code for the current file's older versions.
+        """
         if not task_doc:
             return ""
         if not task_doc.content:
             task_doc = project_repo.docs.task.get(filename=task_doc.filename)
         m = json.loads(task_doc.content)
-        code_filenames = m.get("Task list", [])
+        code_filenames = m.get(TASK_LIST.key, []) if mode == "normal" else m.get(REFINED_TASK_LIST.key, [])
         codes = []
         src_file_repo = project_repo.srcs
-        for filename in code_filenames:
-            if filename == exclude:
-                continue
-            doc = await src_file_repo.get(filename=filename)
-            if not doc:
-                continue
-            codes.append(f"----- {filename}\n" + doc.content)
+
+        if mode == "incremental":
```

Comment:
Define these literal key words in `const.py`

Reply:
remove parameter mode in get_codes function. Use self.config.inc to judge condition